### PR TITLE
Plotting tools to visualize the accuracy of a class at a training set size

### PR DESCRIPTION
**Pull request by akhilaananthram**
Now experiment.py will create two graphs.  The first is a bunch of subplots for each training set size with the accuracy of a certain class and its variance.  The second is a combination of those subplots all in one graph.

@BoltzmannBrain

### CODE REVIEW

File: fluent/utils/plotting.py [hunk 1]
```diff
@@ -117,16 +120,98 @@ def confusionMatrix(self, data, normalize=True):
...
+  def plotCategoryAccuracies(self, trialAccuracies, trainSize):
     """
+    Shows the accuracy for the categories at a certain training size
+    @param trialAccuracies    (dict)    maps a trial size to a dictionary that
+                                        maps a category to a list of accuracies
+    @param trainSize          (list)    size of training set for each trial
     """
-    ## TODO
+    sizes = sorted(set(trainSize))
+    size_sqrt = math.sqrt(len(sizes))
```
Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: You're trying to round up to nearest int? So why not just `int(math.ceil())`?

File: fluent/utils/plotting.py [hunk 2]
```diff
@@ -117,16 +120,98 @@ def confusionMatrix(self, data, normalize=True):
...
-  def plotCategoryAccuracies():
+  def plotCategoryAccuracies(self, trialAccuracies, trainSize):
     """
+    Shows the accuracy for the categories at a certain training size
+    @param trialAccuracies    (dict)    maps a trial size to a dictionary that
+                                        maps a category to a list of accuracies
+    @param trainSize          (list)    size of training set for each trial
     """
-    ## TODO
+    sizes = sorted(set(trainSize))
```
Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: Why make this a set?

File: fluent/utils/plotting.py [hunk 3]
```diff
@@ -117,16 +120,98 @@ def confusionMatrix(self, data, normalize=True):
...
+    size_sqrt = math.sqrt(len(sizes))
+    side = int(size_sqrt)
```
Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: Rename to something like `subplotDimension` to make more clear what is being calculated here; descriptive names make code more readable :smiley:

File: fluent/utils/plotting.py [hunk 4]
```diff
@@ -117,16 +120,98 @@ def confusionMatrix(self, data, normalize=True):
...
+
+    plot_url = py.plot(fig)
+    print "Category Accuracies URL: ", plot_url
+
+
+  def plotCummulativeAccuracies(self, classificationAccuracies, trainSize):
     """
+    Creates scatter plots that show the accuracy for each category at a
+    certain training size
+    @param classificationAccuracies (dict) maps a category label to a list of
```
Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: Same as previous comment on docstring param.

File: fluent/models/classification_model.py [hunk 5]
```diff
@@ -178,7 +179,7 @@ def evaluateCumulativeResults(self, intermResults):
 
 
   @staticmethod
-  def calculateAccuracy(classifications):   ## TODO: just get numpy arrays passed in?
+  def calculateAccuracy(classifications, label=None):   ## TODO: just get numpy arrays passed in?
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: Rename `label` to something more descriptive of its actual function, and please add this param to the docstring.

Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: Now that there is more functionality to this method, the docstring should be more detailed.

File: fluent/models/classification_model.py [hunk 6]
```diff
@@ -119,7 +119,8 @@ def winningLabels(labels, n=3):
   def calculateClassificationResults(self, classifications):  ## TODO: plot
     """Calculate the classification accuracy for each category.
```
Comment (BoltzmannBrain) [inline fluent/models/classification_model.py]: Please add `@return...` info to this docstring.

File: fluent/experiments/runner.py [hunk 7]
```diff
@@ -174,7 +176,28 @@ def calculateResults(self):
       for i in xrange(len(self.trainSize))]
 
     self.model.printFinalReport(self.trainSize, [r[0] for r in resultCalcs])
-    ## TODO: plot accuracies; see Model base class
+
+    # In case there are multiple trials of the same size
+    # trialSize -> (category -> list of accuracies)
+    trialAccuracies = defaultdict(lambda : defaultdict(lambda: numpy.ndarray(0)))
```
Comment (BoltzmannBrain) [inline fluent/experiments/runner.py]: Nice use of `defaultdict`, but please delete the space after the first `lambda`.

File: fluent/experiments/runner.py [hunk 8]
```diff
@@ -174,7 +176,28 @@ def calculateResults(self):
...
+    # Need the accuracies to be ordered for the graph
+    trials = sorted(set(self.trainSize))
+    # category -> list of list of accuracies
+    classificationAccuracies = defaultdict(list)
+    for trial in trials:
+      accuracies = trialAccuracies[trial]
+      for label, acc in accuracies.iteritems():
+        classificationAccuracies[label].append(acc)
+
+    plotter = PlotNLP()
```
Comment (BoltzmannBrain) [inline fluent/experiments/runner.py]: Should only plot when `model.plot = True`

**Comment by BoltzmannBrain** [comment]
@akamlani I left some comments. Please add some simple integration tests to validate we get the expected accuracies.

File: fluent/utils/plotting.py (hunk 2 above)
Comment (akhilaananthram) [inline fluent/utils/plotting.py]: I made that a set in case someone sets trainSize = [2, 2, 2] or anything where a number is repeated.  This way I don't create a separate data point for each trial if they are the same training size.  Do you want separate data points?

Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: We should plot one subplot per trial. So if `trainSize = [2,2,2]` we would have three subplots. A possible experiment where `trainSize` would have repeat values could be several iterations of the same training size where we vary params (e.g. text preprocessing) between each trials.

Comment (akhilaananthram) [inline fluent/utils/plotting.py]: So no variation on the plot then?

Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: Sorry I was confusing this plot with another. Disregard my previous comment.

**Comment by akhilaananthram** [comment]
So you want an integration test to test calculateClassificationResults? not a unit test?

File: fluent/utils/plotting.py [hunk 9]
```diff
@@ -117,16 +120,97 @@ def confusionMatrix(self, data, normalize=True):
     print "Confusion matrix URL: ", plot_url
 
 
-  def plotCategoryAccuracies():
+  def plotCategoryAccuracies(self, trialAccuracies, trainSize):
     """
+    Shows the accuracy for the categories at a certain training size
+    @param trialAccuracies    (dict)    maps a trial size to a dictionary that
```
Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: Please edit this docstring to more clearly explain the param.

Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: Please edit this docstring to more clearly explain the param.

File: fluent/utils/plotting.py [hunk 10]
```diff
@@ -117,16 +120,97 @@ def confusionMatrix(self, data, normalize=True):
...
+
+    plot_url = py.plot(fig)
+    print "Category Accuracies URL: ", plot_url
+
+
+  def plotCummulativeAccuracies(self, classificationAccuracies, trainSize):
     """
+    Creates scatter plots that show the accuracy for each category at a
+    certain training size
+    @param classificationAccuracies (dict) maps a category label to a list of
```
Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: Please edit this docstring to more clearly explain the param.

**Comment by BoltzmannBrain** [comment]
@akhilaananthram I have in mind end-to-end tests to validate both that the experiment outputs the expected results, and that the plots reflect these accurately. However this is beyond the scope of this PR. Let's touch base Monday to discuss how we test this.

For this PR, please revise such that the plotly figures are labeled with category names (e.g. "Management"), not indices. Use `Runner.labelRefs`.

File: fluent/experiments/runner.py [hunk 11]
```diff
@@ -174,7 +176,29 @@ def calculateResults(self):
       for i in xrange(len(self.trainSize))]
 
     self.model.printFinalReport(self.trainSize, [r[0] for r in resultCalcs])
-    ## TODO: plot accuracies; see Model base class
+
+    if self.model.plot:
+      # In case there are multiple trials of the same size
+      # trialSize -> (category -> list of accuracies)
+      trialAccuracies = defaultdict(lambda: defaultdict(lambda: numpy.ndarray(0)))
```
Comment (BoltzmannBrain) [inline fluent/experiments/runner.py]: Please break lines at 80 characters.

File: fluent/experiments/runner.py [hunk 12]
```diff
@@ -174,7 +176,29 @@ def calculateResults(self):
       for i in xrange(len(self.trainSize))]
 
     self.model.printFinalReport(self.trainSize, [r[0] for r in resultCalcs])
-    ## TODO: plot accuracies; see Model base class
+
+    if self.model.plot:
```
Comment (BoltzmannBrain) [inline fluent/experiments/runner.py]: This is the wrong plot variable. `model.plot` is for plotting the confusion matrix in `ClassificationModel`. We want `plot` to be a `Runner` member variable, that can be set with the other experiment.py arguments; make False the default.

**Comment by akhilaananthram** [comment]
@BoltzmannBrain responded to your comments

File: fluent/utils/plotting.py [hunk 13]
```diff
@@ -22,9 +22,12 @@
...
 """
 
+import math
+import numpy
 import os
 import pandas as pd
 import plotly.plotly as py
+import plotly.tools as tls
 
 from plotly.graph_objs import *
```
Comment (BoltzmannBrain) [inline fluent/utils/plotting.py]: Would be better to import the specific objects, not `*`; explicit is better than implicit :smile:

**Comment by akhilaananthram** [comment]
@BoltzmannBrain: I added the specific objects to the import

**Comment by BoltzmannBrain** [comment]
@akhilaananthram please fix the follfowing:
- center the Accuracy and Category Label axis labels
- set the y-axis max to > 1 (currently, points plotted at 1.0 accuracy get cut off)
- fix subplotting so empty plots don't show up
- subplot x-axis should show all labels
[This example](https://plot.ly/~alavin/1018/accuracies-for-category-by-training-size/) shows these issues.

**Comment by akhilaananthram** [comment]
https://plot.ly/~asa225/1581/accuracies-for-category-by-training-size/

**Comment by BoltzmannBrain** [comment]
I'm still not seeing all the category labels on the x-axis:
- https://plot.ly/~alavin/1047/accuracies-for-category-by-training-size/
- https://plot.ly/~alavin/1053/accuracies-for-category-by-training-size/

Nor am I seeing the variance bars...
@akhilaananthram it's important to test your code thoroughly -- i.e. running many examples and corner cases -- to see where it breaks. The code reviewer shouldn't have to do the debugging for you...

**Comment by akhilaananthram** [comment]
There are missing category labels on the x-axis for some of the subplots because at that training size, there were no test instances for that category.  That's why there are missing points for the plots with larger training sizes.

What did you use as the trainSize that resulted in no variance? The only time I get no variance is when I only have one instance of a train size, i.e., all the train sizes are unique.

**Comment by BoltzmannBrain** [comment]
Closing for #94
